sound/soc/sof/ipc4-topology.c

```diff
@@ -1652,24 +1672,9 @@ static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
 		cfg += sizeof(struct sof_ipc4_audio_format);
 	}
 
-	for (i = 0; i < widget->num_kcontrols; i++) {
-		kc = &widget->kcontrol_news[i];
-
-		/* payload uses byte kcontrol */
-		if (widget->dobj.widget.kcontrol_type[i] != SND_SOC_TPLG_TYPE_BYTES)
-			continue;
-		sbe = (struct soc_bytes_ext *)kc->private_value;
-		scontrol = sbe->dobj.private;
-		if (!scontrol)
-			continue;
-		control_data = scontrol->ipc_control_data;
-		if (control_data->data->blob_type != SOF_IPC4_MOD_INIT_INSTANCE)
-			continue;
-
-		data = (void *)control_data->data->data;
-		memcpy(cfg, data, control_data->data->size);
-		break;
-	}
+	control_data = sof_ipc4_get_widget_init_byte_kcontrol(swidget);
+	if (control_data)
+		memcpy(cfg, control_data->data->data, control_data->data->size);
```

Comment:
can we somehow avoid looping through the controls again to copy this data during prepare?

Reply:
I think we can introduce a member in swidget `struct sof_ipc4_control_data *cdata`, and do something like this:
```
if (cdata)
    return cdata;

// code to loop the controls and assign cdata.
```
comments?

Reply:
No please no ipc4 specific members in swidget. What you can do instead is copy the data to process config data during setup and set the size to the size of the data. Then during prepare, push the control data out by the rest of the stuff that needs to precede it

Reply:
I mean process module structure, but failed to update the comment above finally.